File: day2.hs
import System.IO 

main :: IO ()
main = do
    i <- readFile "input2.txt"
    print $ go (processRawInput i) (0,0,0)

processRawInput :: String -> [[String]]
processRawInput rawInput = map words (lines rawInput)

go :: [[String]] -> (Int, Int, Int) -> (Int, Int)
go [] (dist, depth, aim) = (dist * aim, dist * depth)
go inst coords = go (tail inst) (doInst (head inst) coords)

doInst :: [String] -> (Int, Int, Int) -> (Int, Int, Int)
doInst ["forward", x] (dist, depth, aim) = (dist + read x, depth + aim * read x, aim)
doInst ["down", x] (dist, depth, aim)    = (dist, depth, aim + read x)
doInst ["up", x] (dist, depth, aim)      = (dist, depth, aim - read x)

File: day2.py
from aoc_timer import aoc_timer


def go(inst, part):
    horiz, depth, aim = 0, 0, 0
    for x, n in inst:
        if part == 1:
            match x:
                case "forward":
                    horiz += int(n)
                case "down":
                    depth += int(n)
                case "up":
                    depth -= int(n)
        if part == 2:
            match x:
                case "forward":
                    horiz += int(n)
                    depth += int(n)*aim
                case "down":
                    aim += int(n)
                case "up":
                    aim -= int(n)
    return(horiz * depth)

def part1(inst):
    return go(inst, part = 1)

def part2(inst):
    return go(inst, part = 2)

def main():
    f = open("input2.txt")
    inst = [x.split() for x in f.read().split("\n")[:-1]]
    print(aoc_timer(part1, [inst], "Part 1: "))
    print(aoc_timer(part2, [inst], "Part 2: "))

if __name__ == "__main__":
    main()